Source/Plugin.BLE/Android/Device.cs

```diff
@@ -262,58 +263,68 @@ public static List<AdvertisementRecord> ParseScanRecord(byte[] scanRecord)
-                    case AdvertisementRecordType.UuidsIncomple16Bit:
-                    case AdvertisementRecordType.UuidsIncomplete128Bit:
-                        Array.Reverse(data);
+                    if (!Enum.IsDefined(typeof(AdvertisementRecordType), type))
+                    {
+                        Trace.Message("Advertisment record type not defined: {0}", type);
                         break;
-                }
-                var record = new AdvertisementRecord((AdvertisementRecordType)type, data);
+                    }
 
-                Trace.Message(record.ToString());
+                    //data length is length -1 because type takes the first byte
+                    byte[] data = new byte[length - 1];
+                    Array.Copy(scanRecord, index + 1, data, 0, length - 1);
 
-                records.Add(record);
+                    // don't forget that data is little endian so reverse
+                    // Supplement to Bluetooth Core Specification 1
+                    // NOTE: all relevant devices are already little endian, so this is not necessary for any type except UUIDs
+                    //var record = new AdvertisementRecord((AdvertisementRecordType)type, data.Reverse().ToArray());
 
-                //Advance
-                index += length;
+                    switch ((AdvertisementRecordType)type)
+                    {
+                        case AdvertisementRecordType.ServiceDataUuid32Bit:
+                        case AdvertisementRecordType.SsUuids128Bit:
+                        case AdvertisementRecordType.SsUuids16Bit:
+                        case AdvertisementRecordType.SsUuids32Bit:
+                        case AdvertisementRecordType.UuidCom32Bit:
+                        case AdvertisementRecordType.UuidsComplete128Bit:
+                        case AdvertisementRecordType.UuidsComplete16Bit:
+                        case AdvertisementRecordType.UuidsIncomple16Bit:
+                        case AdvertisementRecordType.UuidsIncomplete128Bit:
+                            Array.Reverse(data);
+                            break;
+                    }
+                    var record = new AdvertisementRecord((AdvertisementRecordType)type, data);
+
+                    Trace.Message(record.ToString());
+
+                    records.Add(record);
+
+                    //Advance
+                    index += length;
+                }
+                return records;
+            }
+            catch(Exception)
+            {
+                Trace.Message($"Filed to parse advertisement data: {scanRecord.ToHexString()}");
+                //There may be a situation where scanRecord contains incorrect data.
+                //We return an empty list, even if some of the data is recognized,
+                //because if the scanRecord is invalid, then we cannot be sure of the data.
+                return new List<AdvertisementRecord>();
```

Comment:
Maybe this should better be `return records`? If an advertisement contains multiple records, and the failure occurs only in a later one, we could at least return the earlier ones, which were ok. Right?

Reply:
I did so at the beginning, but then I thought that we could not be sure that the package was not damaged at the beginning or in the middle.
That is, if we extract the middle from the data "02-01-1A-02-0A-08-0C-FF-4C-00-10-07-35-1F-69-21-B3-32-08", for example, starting from index 5 (02-01-1A-02-0A-4C-00-10-07-35-1F-69-21-B3-32- 08), then we will get incorrect data already inside the recognized records.
If the package is in the wrong format, then in theory it can be broken anywhere.
However, I can rewrite the code to return post values.

Reply:
In the example I just gave, there will not be an exception, since the loop will be interrupted at the line
~~~
if (!Enum.IsDefined(typeof(AdvertisementRecordType), type))
{
    Debug.WriteLine("Advertisment record type not defined: {0}", type);
    break;
}
~~~
which is also probably not correct, since it will contain bad data?
Perhaps in this place, too, you need to return an empty array?

Reply:
> I did so at the beginning, but then I thought that we could not be sure that the package was not damaged at the beginning or in the middle.

Well, we never have any guarantee that the data is "correct". We just keep parsing as long as it's plausible.


> However, I can rewrite the code to return recognized messages.

Yeah, I think that's the best we can do. One could simply pull the `return records` statement out of the `try` block and put it after the `catch` block (which should then not contain a `return` statement itself).

Reply:
> In the example I just gave, there will not be an exception, since the loop will be interrupted at the line
> 
> ```
> if (!Enum.IsDefined(typeof(AdvertisementRecordType), type))
> {
>     Debug.WriteLine("Advertisment record type not defined: {0}", type);
>     break;
> }
> ```
> 
> which is also probably not correct, since it will contain bad data. Perhaps in this place, too, you need to return an empty array?

No, IMHO we should parse as much as possible, and stop only once there is an obvious problem. I don't think it's a very likely scenario that the beginning of a record is missing. I assume that most 'bad' advertisements would be sent out by 'buggy' devices that simply transmit non-standard advertisements (intentionally or not).

Reply:
I think you have more experience with this. On this you are right